### DISCUSSION TITLE
Restart Neovim without restarting Neovide

### DISCUSSION THREAD

**Discussion by chrisgrieser** [neovide/neovide#1713] [Q&A] [answered]
For some config changes, it is necessary to restart neovim instead of just re-sourcing the config files. 

While with lazy loading Neovim startup time is incredibly fast, a fair amount of time seems to be taken up by restarting Neovide as well in the process. 

is it possible to restart neovim without restarting Neovide?

**Comment by duckbrain** [comment]
Prior to neovide, I used a shell script to wrap neovim and relaunch it if it exited with a specific status code. Then I made a mapping in nvim to call `os.exit(62)` to trigger the restart.

I attempted to reuse this script using `--neovim-bin`. but it appears that once the headless session disconnects, neovide doesn't reconnect it.  It's also possible neovide doesn't like the state it's left in on the `os.exit`.

**Reply by fredizzimo** [reply]
You should be able to do the same with Neovide, the exit code from nvim is forwarded as the exit code of Neovide. You just need to change your script to launch Neovide instead of nvim. Make sure to use the default `--no-fork`, otherwise it's not possible to get any exit codes.

**Reply by duckbrain** [reply]
Thanks, that sounds good. I'll give that a shot.

I was also trying to avoid losing the window, since that causes layout changes in i3, but getting a form of easy restart will be nice.

**Reply by unphased** [reply]
@duckbrain did it work for you? I am currently (but hoping not to in the future) using Raycast Zoxide Git Projects extension to launch Neovide... It looks like I will have to make a .app wrapper around such a shell script to gain this ability, which is going to be awkward but should definitely work. 

I would be happy with this setup if it weren't for raycast hanging and refusing to open the window if i haven't used this in a while (forcing me to manually launch raycast from spotlight to wake it up), which defeats its entire purpose.

**Reply by duckbrain** [reply]
Yep. The exit code passes through just fine.

In case it's helpful, here is my wrapper script. I ended up changing it to a sentinel file so nvim could shut itself down nicely.

```shell
#!/bin/bash

mkdir -p /tmp/dvim/
export DVIM_SESSION=/tmp/dvim/$$!

while true; do
  neovide --no-fork "$@"
  status="$?"

  if [[ -f "$DVIM_SESSION" ]]; then
    export DVIM_RESTARTED=1
    rm "$DVIM_SESSION"
  else
    exit $status
  fi
done
```

And the nvim binding

```lua
        ["<Leader>R"] = {
          function()
            if not vim.env.DVIM_SESSION then
              vim.notify "This is a normal vim session; unable to restart."
              return
            end

            local file = io.open(vim.env.DVIM_SESSION, "w")
            if file then file:close() end
            vim.cmd.wall()
            vim.cmd.quit()
          end,
          desc = "Restart in-place",
        },
```

**Reply by chrisgrieser** [reply]
If it is about restarting neovide (and not restarting nvim without restarting neovide, as the original topic of this thread): 

For that, I came up with this function that does the trick. It's self-contained, i.e., no need for a wrapper or external script, just add the keymap to your config.

```lua
vim.keymap.set("n", "<leader>rr", function ()
	assert(vim.g.neovide, "Requires neovide.")
	local script = [=[
		while pgrep -xq "neovide" ; do
			sleep 0.05
			i=$((i+1)) ; [[ $i -gt 50 ]] && return 1 # timeout
		done
		sleep 0.1
		[[ "$OSTYPE" =~ "darwin" ]] && open -a "neovide" || neovide # on macOS, use `open -a`
	]=]
	vim.system({ "zsh", "-c", script }, { detach = true }) -- detach to run after nvim quit
	vim.defer_fn(vim.cmd.wqall, 1)
end, { desc = "Restart Neovide" })
```

**Reply by unphased** [reply]
Thanks. @chrisgrieser At first glance your script is going to launch an async child zsh that loops waiting for any and all neovides to exit before relaunching a neovide. My requirements are much more sophisticated than that, though I imagine the same approach could at least in theory still work for my needs. 

I typically have between 5 and 15 (lol) neovide *instances* open. The idea would be to have a way to force the current neovide to exit and then restart on its same working directory. when neovim starts it loads the correct project session based on the working dir so all editor state I happen to care about is handled by the session manager. The biggest issue is probably about how to `pgrep` out the neovide that is on the correct working dir... 

@duckbrain This approach looks simple and robust to me. Neovide always running under a shell wrapper seems ugly but it's a highly practical approach and I see no issues with it. I especially like that pollution from stale sentinel files will be autocleaned on account of landing inside /tmp.

**Reply by unphased** [reply]
I'd like to be able to just restart neovim while keeping the neovide instance intact, but this seems a bit too advanced to attempt in the current state of affairs!

**Reply by chrisgrieser** [reply]
>  I typically have between 5 and 15 (lol) neovide instances open. 

lol, might be worth considering whether that part of your workflow could be improved. I only ever use one neovide instrance by using an autocmd with and `vim.fs.root` to dynamically change cwd depending on the buffer I am switching to.

**Reply by chrisgrieser** [reply]
> I'd like to be able to just restart neovim while keeping the neovide instance intact

yeah, that was (and still is) also my original goal. Maybe nvim 0.12's upcoming `:restart` will provide some method to do this properly with neovide. https://neovim.io/doc/user/gui.html#%3Arestart

**Reply by unphased** [reply]
Yeah there are some exciting stuff coming so i hope i can switch to having a single neovide instance with multiple windows. 

There's not really an issue with having multiple neovide and neovim instances open. previously it would run up RAM usage insanely, but that's largely taken care of once a system is put in place to idle out and kill the LSP servers! Sure I can switch between sessions using a nvim session manager but 1 second saved switching to an instance already open is 1 second saved. Plus sometimes i wanna have two projects' files on screen at the same time.

**Reply by fredizzimo** [reply]
`:restart` would be the correct approach, but it needs support on the Neoivde side as well.

There's a restart event that needs to be handled
> ["restart", progpath, argv]
	[:restart](https://neovim.io/doc/user/gui.html#%3Arestart) command has been used and the Nvim server is about to exit.
	The UI should wait for the server to exit, and then start a new server
	using progpath as the full path to the Nvim executable [v:progpath](https://neovim.io/doc/user/vvars.html#v%3Aprogpath) and
	argv as its arguments [v:argv](https://neovim.io/doc/user/vvars.html#v%3Aargv), and reattach to the new server.
	Note: [--embed](https://neovim.io/doc/user/starting.html#--embed) and [--headless](https://neovim.io/doc/user/starting.html#--headless) are excluded from argv, and the client
	should decide itself whether to add either flag.

**Reply by falcucci** [reply, marked as answer]
@chrisgrieser @unphased this is on my to-do list already. will do when I am back from ny.

**Reply by chrisgrieser** [reply]
Awesome, looking forward to it. And thanks!

**Comment by fredizzimo** [comment]
The refactoring that's taking place in order to support the multigrid will make this much easier to support in the future. So this will probably be implemented sometime after that is done.
